src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -21,127 +24,137 @@
+    String[] teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none()).split(",");
+    String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
+
+    // add kind users to keyfactory
+    keyFactory.setKind("users");
+
+    // take the user elements and create a new entity object for it
+    FullEntity userEntity = Entity.newBuilder(keyFactory.newKey(id)).set("imgURL", imgURL).set("name", name)
+        .set("email", email)
+        .set("wishlist", Arrays.asList(wishlist).stream().map(StringValue::of).collect(Collectors.toList()))
+        .set("teachlist", Arrays.asList(teachlist).stream().map(StringValue::of).collect(Collectors.toList())).build();
+
+    // insert that new entity
+    try {
+      datastore.add(userEntity);
+    } catch (Exception DatastoreException) {
+      this.doPut(request, response);
+      return;
     }
+    response.sendRedirect("index.html");
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException { // update
+
+    // get form information
+    String NOTHING = Jsoup.clean(request.getParameter("NOTHING"), Whitelist.none());
+    String imgURL = Jsoup.clean(request.getParameter("imgURL"), Whitelist.none());
+    String name = Jsoup.clean(request.getParameter("name"), Whitelist.none());
+    String email = Jsoup.clean(request.getParameter("email"), Whitelist.none());
+    String wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none());
+    String teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none());
+    String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
+
+    // turn wishlist into a storable product for datastore
+    String[] wishlistList = wishlist.split(",");
+    List<Value<?>> wishlistValueList = new ArrayList<Value<?>>();
+
+    for (String value : wishlistList) {
+      StringValue tempStrValue = new StringValue(value);
+      wishlistValueList.add(tempStrValue);
+    }
+
+    // turn teachlist into a storable product for datastore
+    String[] teachlistList = teachlist.split(",");
+    List<Value<?>> teachlistValueList = new ArrayList<Value<?>>();
+
+    for (String value : teachlistList) {
+      StringValue tempStrValue = new StringValue(value);
+      teachlistValueList.add(tempStrValue);
+    }
+
+    // access the original entity
+    keyFactory.setKind("users");
+    Key userKey = keyFactory.newKey(id);
+    Entity user = datastore.get(userKey);
+
+    if (wishlist.equals(NOTHING)) {
+      wishlistValueList = user.getList("wishlist");
+    }
+    if (teachlist.equals(NOTHING)) {
+      teachlistValueList = user.getList("teachlist");
+    }
+
+    // take the user elements and create a new updated entity object for it
+    FullEntity userUpdatedEntity = Entity.newBuilder(user).set("imgURL", imgURL).set("name", name).set("email", email)
+        .set("wishlist", wishlistValueList).set("teachlist", teachlistValueList).build();
```

Comment:
With the suggested refactor from the above comment, you can simplify everything above line this to:

```java
FullEntity userEntity = parseUserEntity(request);
```

And then `userEntity` in the next line.